source/extensions/filters/http/cache/hazelcast_http_cache/hazelcast_context.cc

```diff
@@ -0,0 +1,511 @@
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace HazelcastHttpCache {
+
+using Envoy::Protobuf::util::MessageDifferencer;
+
+HazelcastLookupContextBase::HazelcastLookupContextBase(HazelcastCache& cache,
+                                                       LookupRequest&& request)
+    : hz_cache_(cache), lookup_request_(std::move(request)) {
+  createVariantKey(lookup_request_.key());
+  variant_hash_key_ = stableHashKey(lookup_request_.key());
+}
+
+void HazelcastLookupContextBase::getTrailers(LookupTrailersCallback&&) {
+  // TODO(enozcan): Support trailers when implemented on the filter side.
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+void HazelcastLookupContextBase::handleLookupFailure(absl::string_view message,
+                                                     const LookupHeadersCallback& cb,
+                                                     bool warn_log) {
+  if (warn_log) {
+    ENVOY_LOG(warn, "{}", message);
+  } else {
+    ENVOY_LOG(debug, "{}", message);
+  }
+  abort_insertion_ = true;
+  cb(LookupResult{});
+}
+
+void HazelcastLookupContextBase::createVariantKey(Key& raw_key) {
+  ASSERT(raw_key.custom_fields_size() == 0);
+  ASSERT(raw_key.custom_ints_size() == 0); // Key must be pure.
+  if (lookup_request_.vary_headers().size() == 0) {
+    return;
+  }
+  std::vector<std::pair<std::string, std::string>> header_strings;
+
+  for (const Http::HeaderEntry& header : lookup_request_.vary_headers()) {
+    header_strings.push_back(std::make_pair(std::string(header.key().getStringView()),
+                                            std::string(header.value().getStringView())));
+  }
+
+  // Different order of headers causes different hash keys even if their both key and value
+  // are the same. That is, the following two header lists will cause different hashes for
+  // the same response and hence they are sorted before insertion.
+  //
+  // { {"User-Agent", "desktop"}, {"Accept-Encoding","gzip"} }
+  // { {"Accept-Encoding","gzip"}, {"User-Agent", "desktop"} }
+
+  std::sort(header_strings.begin(), header_strings.end(), [](auto& left, auto& right) -> bool {
+    // Per https://tools.ietf.org/html/rfc2616#section-4.2 if two different header entries
+    // have the same field-name, then their order should not change. For distinct field-named
+    // headers the order is not significant but sorted alphabetically here to get the same hash
+    // for the same headers.
+    return left.first == right.first ? false : left.first < right.first;
+  });
+
+  // stableHashKey now creates variant hash for the key since its custom_fields are like:
+  // [ "Accept-Encoding", "gzip", "User-Agent", "desktop"]
+  for (auto& header : header_strings) {
+    raw_key.add_custom_fields(std::move(header.first));
+    raw_key.add_custom_fields(std::move(header.second));
+  }
+
+  // TODO(enozcan): Ensure the generation of the same key for the same response independent
```

Comment:
I could not decide if parsing all vary headers for a response and then sorting them would be a right approach.